src/providers/aws.rs

```diff
@@ -0,0 +1,887 @@
+    pub fn region(self, region: Region) -> Setup<NoAmi> {
+        Setup {
+            region,
+            instance_type: self.instance_type,
+            ami: None,
+            setup: self.setup,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+/// AWS EC2 spot instance launcher.
+///
+/// Each individual region is handled by `RegionLauncher`.
+#[derive(Educe)]
+#[educe(Debug)]
+pub struct Launcher<P = DefaultCredentialsProvider> {
+    #[educe(Debug(ignore))]
+    credential_provider: Box<dyn Fn() -> Result<P, Error>>,
+    max_instance_duration_hours: usize,
+    regions: HashMap<<Setup<YesAmi> as super::MachineSetup>::Region, RegionLauncher>,
+}
+
+impl Default for Launcher {
+    fn default() -> Self {
+        Launcher {
+            credential_provider: Box::new(|| Ok(DefaultCredentialsProvider::new()?)),
+            max_instance_duration_hours: 6,
+            regions: Default::default(),
+        }
+    }
+}
+
+impl<P> Launcher<P> {
+    /// `Launcher` uses [defined duration](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-requests.html#fixed-duration-spot-instances)
+    /// instances.
+    ///
+    /// The lifetime of such instances must be declared in advance (1-6 hours).
+    /// This method thus clamps `t` to be between 1 and 6.
+    ///
+    /// By default, we use 6 hours (the maximum).
+    pub fn set_max_instance_duration(&mut self, t: usize) -> &mut Self {
+        let t = std::cmp::min(t, 6);
+        let t = std::cmp::max(t, 1);
+        self.max_instance_duration_hours = t;
+        self
+    }
+
+    /// Set the credential provider used to authenticate to EC2.
+    ///
+    /// The provided function is called once for each region, and is expected to produce a
+    /// [`P: ProvideAwsCredentials`](https://docs.rs/rusoto_core/0.40.0/rusoto_core/trait.ProvideAwsCredentials.html)
+    /// that gives access to the region in question.
+    pub fn with_credentials<P2>(self, f: impl Fn() -> Result<P2, Error> + 'static) -> Launcher<P2> {
+        Launcher {
+            credential_provider: Box::new(f),
+            max_instance_duration_hours: self.max_instance_duration_hours,
+            regions: self.regions,
+        }
+    }
+}
+
+impl<P> Launcher<P>
+where
+    P: ProvideAwsCredentials + Send + Sync + 'static,
+    <P as ProvideAwsCredentials>::Future: Send,
+{
+    fn get_credential_provider(&self) -> Result<P, Error> {
+        (*self.credential_provider)()
+    }
+}
+
+impl<P> super::Launcher for Launcher<P>
+where
+    P: ProvideAwsCredentials + Send + Sync + 'static,
+    <P as ProvideAwsCredentials>::Future: Send,
+{
+    type MachineDescriptor = Setup<YesAmi>;
+
+    fn launch(&mut self, l: super::LaunchDescriptor<Self::MachineDescriptor>) -> Result<(), Error> {
+        let prov = self.get_credential_provider()?;
+        let mut awsregion = RegionLauncher::new(&l.region.to_string(), prov, l.log)?;
```

Comment:
Should the sequence of calls on `RegionLauncher` really just be a method on `RegionLauncher` instead? That makes more sense to me, because that method can then be `pub`, and the various lower-level methods (like `wait_for_spot_instance_requests`) can be non-`pub`.